# 04 데이터 핸들링 - Pandas
------



 ## [1] 판다스 시작 - 파일을 DataFrame으로 로딩, 기본 API
 





* 판다스 모듈 임포트

In [2]:
import pandas as pd

* read_csv(filepath_or_buffer, sep=',',..) : CSV(칼럼은 ','로 구분한 파일 포맷) 파일 포맷 변환을 위한 API이다.

1) read_table()와의 차이는 read_table()은 디폴트 필드 구분 문자는 탭('\t)이다.  


2) read.csv()는 CSV뿐만 아니라 어떤 필드 구분 문자 기반 파일 포맷도 DataFrame으로 변환이 가능하다.


3) read_csv()의 인자인 sep에 해당 구분 문자를 쓰면 된다.  

(탭으로 필드가 구분돼 있다면 read_csv('파일명', sep='\t')으로 쓰고, sep 인자를 생략하면 자동으로 콤마로 할당한다.(sep=','))  
4) Filepath 인자에는 로드하려는 데이터 파일의 경로를 포함한 파일명을 입력한다. 


5) 별다른 파라미터 지정이 없으면 파일의 맨 처음 로우를 칼럼명으로 인지한다.    


* read_fwf()는 Fixed Width, 즉 고정 길이 기반의 칼럼 포맷을 DataFrame으로 로딩하기 위한 API이다.

In [4]:
# C:\Users\chkwon\Data_Handling 디렉터리에 titanic_train.csv 파일을 저장했을 때 로딩하는 방법 
# titanic_df = pd.read_csv(r'C:\Users\chkwon\Data_Handling\titanic_train.csv')

In [9]:
# 주피터 노트북과 titanic_train.csv가 모두 같은 디렉터리에 있을 경우 로딩하는 방법 
titanic_df = pd.read_csv('titanic_train.csv')
print(type(titanic_df))
titanic_df.head()

<class 'pandas.core.frame.DataFrame'>


PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

* 인덱스: 칼럼명이 표시되지 않는 데이터들로 고유의 레코드를 식별하는 역할을 한다.
* DataFrame.head(): DataFrame의 맨 앞에 있는 N개의 로우를 반환한다.(Default는 5개)
* DataFrame.shape: DF의 행과 열을 튜플 형태로 반환한다

In [16]:
print(titanic_df.shape)

(891, 12)


891개의 로우와 12개의 칼럼으로 이루어짐 

* info() 메서드: 총 데이터 건수와 데이터 타입, Null 건수를 알 수 있다.


* describe(): 칼럼별 숫자형 데이터값의 n-percentile 분포도, 평균값, 최댓값, 최솟값을 나타냄
  - 오직 숫자형(int, float 등) 칼럼의 분포도만 조사하며 자동으로 object 타입의 칼럼은 출력에서 제외시킨다.

In [12]:
titanic_df.info()
# object 타입은 문자열 타입으로 생각해도 무방함 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [13]:
titanic_df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

< Pclass 칼럼값이 어떠한 분포로 구성되어 있는지 살펴보기 >
* value_counts(): 해당 칼럼값의 유형과 건수를 확인할 수 있다. 

In [15]:
value_counts = titanic_df['Pclass'].value_counts()
print(value_counts)

3    491
1    216
2    184
Name: Pclass, dtype: int64


Pclass값 3이 491개, 1이 216개, 2가 184개이다. value_counts()는 많은 건수 순서로 정렬되어 값을 반환한다. 

* DataFrame의 [] 연산자 내부에 칼럼명을 입력하면 해당 칼럼에 해당하는 Series 객체를 반환한다.

In [17]:
titanic_pclass = titanic_df['Pclass']
print(type(titanic_pclass))

<class 'pandas.core.series.Series'>


In [18]:
titanic_pclass.head()

0    3
1    1
2    3
3    1
4    3
Name: Pclass, dtype: int64

Series는 Index와 단 하나의 칼럼으로 구성된 데이터 세트이다. 맨 왼쪽은 인덱스이고 오른쪽은 Series의 해당 칼럼의 데이터값이다. 모든 SEries와 DataFrame은 인덱스를 반드시 가진다. value_counts() 메서드는 Series 객체에서만 정의돼 있다. DataFrame은 value_counts() 메서드를 가지고 있지 않다.

In [19]:
value_counts = titanic_df['Pclass'].value_counts()
print(type(value_counts))
print(value_counts)

<class 'pandas.core.series.Series'>
3    491
1    216
2    184
Name: Pclass, dtype: int64


value_counts()가 반환하는 데이터 타입도 Series 객체이다. 왼쪽이 인덱스값이고, 오른쪽이 데이터값이다. 이 인덱스는 0부터 시작하는 것이 아닌 Pclass 칼럼 값 3, 1, 2를 나타내고 있다. 이처럼 인덱스는 고유성이 보장된다면 의미 있는 데이터값 할당도 가능하다. value_counts()는 칼럼 값별 데이터 건수를 반환하므로 고유 칼럼 값을 식별자로 사용할 수 있다. 


## [2] DataFrame과 리스트, 딕셔너리, 넘파이 ndarray 상호 변환
---
기본적으로 DataFrame은 파이썬의 리스트, 딕셔너리 그리고 넘파이 ndarray 등 다양한 데이터로부터 생성될 수 있다. 또한 DataFrame은 반대로 파이썬의 리스트, 딕셔너리 그리고 넘파이 ndarray 등으로 변환될 수 있다. 


### (1) 넘파이 ndarray, 리스트, 딕셔너리를 DataFrame으로 변환하기

* 일반적으로 DataFrame으로 변환 시 칼럼명을 지정해준다.   


* DataFrame은 기본적으로 행과 열을 가지는 2차원 데이터이다. 따라서 2차원 이하의 데이터들만 DF로 변환될 수 있다. 

> **1차원 형태의 리스트와 넘파이 ndarray를 DataFrame으로 변환하기** 

In [21]:
import numpy as np

col_name1 = ['col1']
list1 = [1, 2, 3]
array1 = np.array(list1)
print('array1 shape:', array1.shape)
# 리스트를 이용해 DataFrame 생성
df_list1 = pd.DataFrame(list1, columns=col_name1)
print('1차원 리스트로 만든 DataFrame:\n', df_list1)
# 넘파이 ndarray를 이용해 DataFrame 생성
df_array1 = pd.DataFrame(array1, columns=col_name1)
print('1차원 ndarray로 만든 DataFrame:\n', df_array1)

array1 shape: (3,)
1차원 리스트로 만든 DataFrame:
    col1
0     1
1     2
2     3
1차원 ndarray로 만든 DataFrame:
    col1
0     1
1     2
2     3


1차원 형태의 데이터를 기반으로 DataFrame을 생성하므로 칼럼명이 한 개만 필요하다. 

> **2차원 형태의 데이터를 기반으로 DataFrame을 생성하기**

In [22]:
# 2행 3열 형태의 리스트와 ndarray를 기반으로 DataFrame을 생성하므로 칼럼명은 3개가 필요함
col_name2 = ['col1', 'col2', 'col3']

# 2행x3열 형태의 리스트와 ndarray 생성한 뒤 이를 DataFrame으로 변환 
list2 = [[1, 2, 3],
         [11, 12, 13]]
array2 = np.array(list2)
print('array2 shape:', array2.shape)
df_list2 = pd.DataFrame(list2, columns=col_name2)
print('2차원 리스트로 만든 DataFrame:\n', df_list2)
df_array2 = pd.DataFrame(array2, columns=col_name2)
print('2차원 ndarray로 만든 DataFrame:\n', df_array2)


array2 shape: (2, 3)
2차원 리스트로 만든 DataFrame:
    col1  col2  col3
0     1     2     3
1    11    12    13
2차원 ndarray로 만든 DataFrame:
    col1  col2  col3
0     1     2     3
1    11    12    13


> **딕셔너리를 DataFrame으로 변환하기**

In [23]:
# Key는 문자열 칼럼명으로 매핑, Value는 리스트 형(또는 ndarray) 칼럼 데이터로 매핑
dict = {'col1':[1, 11], 'col2':[2, 22], 'col3':[3, 33]}
df_dict = pd.DataFrame(dict)
print(df_dict)

   col1  col2  col3
0     1     2     3
1    11    22    33


Key 값은 칼럼명, 각 Value는 각 칼럼 데이터로 매핑되었다. 

### (2) DataFrame을 넘파이 ndarray, 리스트, 딕셔너리로 변환하기

* DataFrame을 넘파이 ndarray로 변환하는 것은 DataFrame 객체의 values를 이용해 할 수 있다.

In [24]:
# DataFrame을 ndarray로 변환
array3 = df_dict.values
print(type(array3))
print(array3.shape)
print(array3)

<class 'numpy.ndarray'>
(2, 3)
[[ 1  2  3]
 [11 22 33]]


In [25]:
# DataFrame을 리스트로 변환 - tolist() 메소드
list3 = df_dict.values.tolist()
print(type(list3))
print(list3)

<class 'list'>
[[1, 2, 3], [11, 22, 33]]


In [26]:
# DataFrame을 딕셔너리로 변환 
# to_dict() 메소드 - 인자로 'list'를 입력하면 딕셔너리의 값이 리스트형으로 반환됨 
dict3 = df_dict.to_dict('list')
print(type(dict3))
print(dict3)

<class 'dict'>
{'col1': [1, 11], 'col2': [2, 22], 'col3': [3, 33]}


## [3] DataFrame의 칼럼 데이터 세트 생성과 수정
---
* [] 연산자를 이용해서 할 수 있다.


> **DataFrame의 새로운 칼럼을 생성하기**

In [27]:
# Titanic DF의 새로운 칼럼 Age_0을 추가하고 일괄적으로 0값을 할당하기
titanic_df['Age_0'] = 0
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_0  
0      0         A/5 21171   7.2500   NaN        S      0  
1      0          PC 17599  71.2833   C85        C      0  
2      0  STON/O2. 3101282   7.9250   NaN        S      0

In [28]:
# 기존 칼럼 Series의 데이터를 이용해 새로운 칼럼 Series 만들기 
titanic_df['Age_by_10'] = titanic_df['Age']*10
titanic_df['Family_No'] = titanic_df['SibSp'] + titanic_df['Parch'] + 1
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_0  Age_by_10  \
0      0         A/5 21171   7.2500   NaN        S      0      220.0   
1      0          PC 17599  71.2833   C85        C      0      380.0   
2      0  STON/O2. 3101282   7.9250   NaN        S      0      260.0   

   Family_No  
0          2  
1          2  
2          1

> **DataFrame 내의 기존 칼럼 값을 일괄적으로 업데이트하기**

In [32]:
titanic_df['Age_by_10'] += 100
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_0  Age_by_10  \
0      0         A/5 21171   7.2500   NaN        S      0      320.0   
1      0          PC 17599  71.2833   C85        C      0      480.0   
2      0  STON/O2. 3101282   7.9250   NaN        S      0      360.0   

   Family_No  
0          2  
1          2  
2          1

## [4] DataFrame 데이터 삭제
---
* DataFrame에서 데이터 삭제는 drop() 메서드를 이용한다.


* **DataFrame.drop(labels=None, axis=0, index=None, columns=None, level=None, inplace=False, errors='raise')**  


axis 값에 따라 특정 칼럼 또는 행을 드롭한다. axis=0은 로우 방향 축, axis 1은 칼럼 방향 축이다.  

drop() 메서드에 axis=1을 입력하면 칼럼 축 방향으로 드롭을 수행하므로 지정된 칼럼을 드롭한다.  

drop() 메서드에 axis=0을 입력하면 로우 축 방향으로 드롭을 수행하므로 특정 로우를 드롭한다.   





In [33]:
# Titanic DataFrame에 추가된 'Age_0' 칼럼을 삭제하기
titanic_drop_df = titanic_df.drop('Age_0', axis=1)
titanic_drop_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_by_10  Family_No  
0      0         A/5 21171   7.2500   NaN        S      320.0          2  
1      0          PC 17599  71.2833   C85        C      480.0          2  
2      0  STON/O2. 3101282   7.9250   NaN        S      360.0          1

In [35]:
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_0  Age_by_10  \
0      0         A/5 21171   7.2500   NaN        S      0      320.0   
1      0          PC 17599  71.2833   C85        C      0      480.0   
2      0  STON/O2. 3101282   7.9250   NaN        S      0      360.0   

   Family_No  
0          2  
1          2  
2          1

원본 Titanic DataFrame을 다시 확인해보면 'Age_0' 칼럼이 여전히 존재한다. 그 이유는 앞의 예제코드에서 inplace=False로 설정했기 때문이다. inplace는 디폴트 값이 False이다. 


* inplace=False이면 자기 자신의 DF의 데이터는 삭제하지 않으며, 삭제된 결과 DF를 반환한다. 


* inplace=True로 설정하면 자신의 DataFrame의 데이터를 삭제한다. 


* 여러 개의 칼럼을 삭제하고 싶으면 리스트 형태로 삭제하고자 하는 칼럼 명을 입력해 labels 파라미터로 입력하면 된다. 


In [36]:
# titanic_df의 'Age_0', 'Age_by_10', 'Family_No' 칼럼들을 삭제하기
drop_result = titanic_df.drop(['Age_0', 'Age_by_10', 'Family_No'], axis=1, inplace=True)
print(drop_result)
titanic_df.head(3)

None


PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

유의할 점은 drop()시 inplace=True로 설정하면 반환 값이 None(아무 값도 아님)이 된다.

* inplace=True로 설정한 채로 반환 값을 다시 자신의 DataFrame 객체로 할당하면 안된다. 

In [38]:
# axis=0으로 설정해 index 0, 1, 2(맨 앞 3개 데이터) 로우를 삭제해보기
pd.set_option('display.width', 1000)
pd.set_option('display.max_colwidth', 15)
print('#### before axis 0 drop ####')
print(titanic_df.head(3))
titanic_df.drop([0, 1, 2], axis=0, inplace=True)

print('#### after axis 0 drop ####')
print(titanic_df.head(3))

#### before axis 0 drop ####
   PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked
0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500   NaN        S
1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C
2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250   NaN        S
#### after axis 0 drop ####
   PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch  Ticket     Fare Cabin Embarked
3            4         1       1  Futrelle, M...  female  35.0      1      0  113803  53.1000  C123        S
4            5         0       3  Allen, Mr. ...    male  35.0      0      0  373450   8.0500   NaN        S
5            6         0       3  Moran, Mr. ...    male   NaN      0      0  330877   8.4583   NaN        Q


## [5] Index 객체
---
< titanic_df DataFrame에서 Index 객체를 추출해보기>

In [41]:
# 원본 파일 다시 로딩
titanic_df = pd.read_csv('titanic_train.csv')
# Index 객체 추출
indexes = titanic_df.index
print(indexes)
# Index 객체를 실제 값 array로 변환
print(indexes.values)

RangeIndex(start=0, stop=891, step=1)
[  0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17
  18  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35
  36  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53
  54  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71
  72  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89
  90  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107
 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125
 126 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143
 144 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161
 162 163 164 165 166 167 168 169 170 171 172 173 174 175 176 177 178 179
 180 181 182 183 184 185 186 187 188 189 190 191 192 193 194 195 196 197
 198 199 200 201 202 203 204 205 206 207 208 209 210 211 212 213 214 215
 216 217 218 219 220 221 222 223 224 225 226 227 228 229 230 231 232 233
 234 235 236 

Index 객체는 식별성 데이터를 1차원 array로 가지고 있다. 또한 ndarray와 유사하게 단일 값 반환 및 슬라이싱도 가능하다. 


In [42]:
print(type(indexes.values))
print(indexes.values.shape)
print(indexes[:5].values)
print(indexes.values[:5])
print(indexes[6])

<class 'numpy.ndarray'>
(891,)
[0 1 2 3 4]
[0 1 2 3 4]
6


* 한번 만들어진 DataFrame 및 Series의 Index 객체는 함부로 변경할 수 없다.

In [75]:
# indexes[0] = 5 -> 오류

Series 객체는 Index 객체를 포함하지만 Series 객체에 연산 함수를 적용할 때는 Index는 연산에서 제외된다. Index는 오직 식별용으로만 사용된다. 

In [45]:
series_fare = titanic_df['Fare']
print(series_fare.max())
print(series_fare.sum())
print(sum(series_fare))
print((series_fare + 3).head(3))

512.3292
28693.9493
28693.949299999967
0    10.2500
1    74.2833
2    10.9250
Name: Fare, dtype: float64


* DataFrame 및 Series에 reset_index() 메서드를 수행하면 새롭게 인덱스를 연속 숫자 형으로 할당하며 기존 인덱스는 'index'라는 새로운 칼럼 명으로 추가합니다. 

In [47]:
titanic_reset_df = titanic_df.reset_index(inplace=False)
titanic_reset_df.head(3)

level_0  index  PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked
0        0      0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500   NaN        S
1        1      1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C
2        2      2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250   NaN        S

* reset_index()는 인덱스가 연속된 int 숫자형 데이터가 아닐 경우에 다시 이를 연속 int 숫자형 데이터로 만들 때 주로 사용한다. 


* Series에 reset_index()를 적용하면 Series가 아닌 DataFrame이 반환된다.

In [49]:
print('### before reset_index ###')
value_counts = titanic_df['Pclass'].value_counts()
print(value_counts)
print(type(value_counts))
new_value_counts = value_counts.reset_index(inplace=False)
print('### After reset_index ###')
print(new_value_counts)
print(type(new_value_counts))

### before reset_index ###
3    491
1    216
2    184
Name: Pclass, dtype: int64
<class 'pandas.core.series.Series'>
### After reset_index ###
   index  Pclass
0      3     491
1      1     216
2      2     184
<class 'pandas.core.frame.DataFrame'>


Series에 reset_index()를 적용하면 새롭게 연속 숫자형 인덱스가 만들어지고 기존 인덱스는 'index' 칼럼명으로 추가되면서 DataFrame으로 변환된다.

* reset_index()의 parameter 중 drop=True로 설정하면 기존 인덱스는 새로운 칼럼으로 추가되지 않고 삭제(drop)된다.  

새로운 칼럼이 추가되지 않으므로 그대로 Series로 유지된다.

## [6] 데이터 셀렉션 및 필터링
---


### (1) DataFrame의 [ ] 연산자
넘파이에서 [] 연산자는 행의 위치, 열의 위치, 슬라이싱 범위 등을 지정해 데이터를 가져올 수 있다. 하지만 DataFrame 바로 뒤에 있는 '[]'안에 들어갈 수 있는 것은 칼럼 명 문자(도는 칼럼 명의 리스트 객체), 또는 인덱스로 변환 가능한 표현식이다.  


* DataFrame에 ['칼럼명']으로 칼럼 데이터의 일부만 추출하기

In [50]:
# 단일 칼럼 데이터 추출
print(titanic_df['Pclass'].head(3))
# 여러 칼럼의 데이터 추출
print(titanic_df[['Survived', 'Pclass']].head(3))
# print(titanic_df[0]) -> 오류 발생 

0    3
1    1
2    3
Name: Pclass, dtype: int64
   Survived  Pclass
0         0       3
1         1       1
2         1       3


* 판다스의 인덱스 형태로 변환 가능한 표현식은 [ ] 내에 입력할 수 있다.

In [51]:
# titanic_df의 처음 2개의 데이터 추출하기 위한 슬라이싱
titanic_df[0:2]

index  PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch     Ticket     Fare Cabin Embarked
0      0            1         0       3  Braund, Mr....    male  22.0      1      0  A/5 21171   7.2500   NaN        S
1      1            2         1       1  Cumings, Mr...  female  38.0      1      0   PC 17599  71.2833   C85        C

* 불린 인덱싱 표현도 가능하다.    

< Pclass 칼럼 값이 3인 데이터 3개만 추출하기 >

In [52]:
titanic_df[titanic_df['Pclass'] == 3].head(3)

index  PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket   Fare Cabin Embarked
0      0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171  7.250   NaN        S
2      2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...  7.925   NaN        S
4      4            5         0       3  Allen, Mr. ...    male  35.0      0      0          373450  8.050   NaN        S

* DataFrame 뒤의 [ ] 연산자는 넘파이의 [ ]나 Series의 [ ]와 다르다. 


* DataFrame 바로 뒤의 [ ] 내 입력 값은 칼럼명(또는 칼럼 리스트)을 지정해 칼럼 지정 연산에 사용하거나 불린 인덱스 용도로만 사용해야 한다.


* DataFrame[0:2]와 같은 슬라이싱 연산으로 데이터를 추출하는 방법은 사용하지 않는게 좋다.


### (2) DataFrame ix[ ] 연산자
* ix[]는 ix[0, 'Pclass']와 같이 행 위치 지정으로 인덱스값 0, 열 위치 지정으로 칼럼 명인 'Pclass'를 입력해 원하는 위치의 데이터를 추출할 수 있다. -> 칼럼 명칭 기반 인덱싱


* ix[]의 행에 해당하는 위치 지정은 DataFrame의 인덱스값을 입력해야 한다. 열 위치 지정은 칼럼 명뿐만 아니라 ix[0, 2]와 같이 칼럼 명이 아닌 칼럼의 위치 값 지정도 가능하다. -> 칼럼 위치 기반 인덱싱 

 예) ix[0, 2]는 인덱스 0인 로우의 세 번째 칼럼에 있는 데이터를 의미함


* ix[]는 현재 판다스에서 사라져서 출력할 경우 Deprecation Warning이 출력된다. 

In [74]:
# print(titanic_df.ix[0, 2])
# print(titanic_df.ix[0, 'Pclass'])

### (3) DataFrame iloc[ ] 연산자 - 위치 기반 인덱싱


* iloc[ ]은 위치 기반 인덱싱만 허용하기 때문에 행과 열 값으로 integer 또는 integer형의 슬라이싱, 팬시 리스트 값을 입력해줘야 한다.  

< data_df의 첫 번째 행, 첫 번째 열의 데이터를 추출하기 >

In [68]:
data = {'Name': ['Chulmin', 'Eunkyung', 'Jinwoong', 'Soobeom'],
        'Year': [2011, 2016, 2015, 2015],
        'Gender': ['Male', 'Female', 'Male', 'Male']}
data_df = pd.DataFrame(data, index=['one', 'two', 'three', 'four'])
data_df_reset = data_df.reset_index()
data_df_reset.index = data_df_reset.index+1
print(data_df_reset)

   index      Name  Year  Gender
1    one   Chulmin  2011    Male
2    two  Eunkyung  2016  Female
3  three  Jinwoong  2015    Male
4   four   Soobeom  2015    Male


In [69]:
data_df_reset.iloc[0, 1]

'Chulmin'

### (4) DataFrame loc[ ] 연산자 - 명칭 기반 인덱싱
* loc[ ]는 명칭 기반으로 데이터를 추출한다.  

* 행의 위치는 DataFrame index 값을, 열 위치에는 칼럼 명을 입력한다.  

< 인덱스값이 'one'인 행의 칼럼 명이 'name'인 데이터를 추출해보기 >


In [70]:
print(data_df)
data_df.loc['one', 'Name']

           Name  Year  Gender
one     Chulmin  2011    Male
two    Eunkyung  2016  Female
three  Jinwoong  2015    Male
four    Soobeom  2015    Male


'Chulmin'

In [71]:
data_df_reset.loc[1, 'Name']

'Chulmin'

In [73]:
# 오류 발생 -> 인덱스값이 0인 행이 없음
# data_df_reset.loc[0, 'Name']

* loc[ ]에 슬라이싱 기호 ':'를 적용하면 종료 값-1이 아니라 종료 값까지 포함하는 것을 의미한다.



* 명칭 기반 인덱싱의 특성: 명칭은 숫자 형이 아닐 수 있기 때문에 -1을 할 수가 없다. 

In [78]:
print(data_df, '\n')
print(data_df.iloc[0:1, 0], '\n')
print(data_df.loc['one':'two', 'Name'])

           Name  Year  Gender
one     Chulmin  2011    Male
two    Eunkyung  2016  Female
three  Jinwoong  2015    Male
four    Soobeom  2015    Male 

one    Chulmin
Name: Name, dtype: object 

one     Chulmin
two    Eunkyung
Name: Name, dtype: object


In [79]:
print(data_df_reset, '\n')
print(data_df_reset.loc[1:2, 'Name'])

   index      Name  Year  Gender
1    one   Chulmin  2011    Male
2    two  Eunkyung  2016  Female
3  three  Jinwoong  2015    Male
4   four   Soobeom  2015    Male 

1     Chulmin
2    Eunkyung
Name: Name, dtype: object


loc[1:2, 'name']에서 1:2는 DataFrame의 인덱스값에 따른 명칭 기반 인덱싱이므로 인덱스가 1, 2인 2개의 데이터를 반환한다.

* **정리**  

1) DataFrame의 인덱스나 칼럼명으로 데이터에 접근하는 것은 명칭 기반 인덱싱이다. 0부터 시작하는 행, 열의 위치 좌표에만 의존하는 것이 위치 기반 인덱싱이다. 

2) ix[ ]는 명칭 기반 인덱싱과 위치 기반 인덱싱을 모두 적용할 수 있다. DataFrame의 인덱스가 숫자 형일 경우 행 위치에 오는 숫자는 위치 기반 인덱싱이 아니라 명칭 기반 인덱싱의 DataFrame 인덱스를 가리킨다.

3) iloc[ ]는 위치 기반 인덱싱만 가능하다. 행과 열의 위치 값으로 정수형 값을 지정해 원하는 데이터를 반환한다. 

4) loc[ ]는 명칭 기반 인덱싱만 가능하다.행 위치에 DataFrame 인덱스가 오며, 열 위치에는 칼럼 명을 지정해 원하는 데이터를 반환한다. 

5) 명칭 기반 인덱싱에서 슬라이싱을 '시작점:종료점'으로 지정할 때 시작점에서 종료점을 포함한 위치에 있는 데이터를 반환한다.

### (5) 불린 인덱싱

< 승객 중 나이(Age)가 60세 이상인 데이터를 추출해보기 >

In [80]:
titanic_df = pd.read_csv('titanic_train.csv')
titanic_boolean = titanic_df[titanic_df['Age'] > 60]
print(type(titanic_boolean))
print(titanic_boolean)

<class 'pandas.core.frame.DataFrame'>
     PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch       Ticket      Fare        Cabin Embarked
33            34         0       2  Wheadon, Mr...    male  66.0      0      0   C.A. 24579   10.5000          NaN        S
54            55         0       1  Ostby, Mr. ...    male  65.0      0      1       113509   61.9792          B30        C
96            97         0       1  Goldschmidt...    male  71.0      0      0     PC 17754   34.6542           A5        C
116          117         0       3  Connors, Mr...    male  70.5      0      0       370369    7.7500          NaN        Q
170          171         0       1  Van der hoe...    male  61.0      0      0       111240   33.5000          B19        S
252          253         0       1  Stead, Mr. ...    male  62.0      0      0       113514   26.5500          C87        S
275          276         1       1  Andrews, Mi...  female  63.0      1      0        13502   

< 60세 이상인 승객의 나이와 이름만 추출해보기 >
* [] 연산자에 칼럼 명을 입력한다. 칼럼이 두 개 이상이므로 [[ ]]를 사용한다.

In [81]:
titanic_df[titanic_df['Age'] > 60][['Name', 'Age']].head(3)

Name   Age
33  Wheadon, Mr...  66.0
54  Ostby, Mr. ...  65.0
96  Goldschmidt...  71.0

In [82]:
# loc[]을 사용하기
titanic_df.loc[titanic_df['Age'] > 60, ['Name', 'Age']].head(3)

Name   Age
33  Wheadon, Mr...  66.0
54  Ostby, Mr. ...  65.0
96  Goldschmidt...  71.0

* and 조건일 때는 &

* or 조건일 때는 |

* Not 조건일 때는 ~   



< 나이가 60세 이상이고, 선실 등급이 1등급이며, 성별이 여성인 승객을 추출해보기 >

* 개별 조건은 ( )로 묶고, 복합 조건 연산자를 사용하면 된다.

In [83]:
titanic_df[ (titanic_df['Age'] > 60) & (titanic_df['Pclass'] == 1) &
          (titanic_df['Sex'] == 'female')]

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch  Ticket     Fare Cabin Embarked
275          276         1       1  Andrews, Mi...  female  63.0      1      0   13502  77.9583    D7        S
829          830         1       1  Stone, Mrs....  female  62.0      0      0  113572  80.0000   B28      NaN

## [7] 정렬, Aggregation 함수, GroupBy 적용
---

### (1) DataFrame, Series의 정렬 - sort_values()

* sort_values()의 주요 입력 파라미터는by, ascending, inplace이다.
 * by로 특정 칼럼을 입력하면 해당 칼럼으로 정렬을 수행한다.
 * ascending=True로 설정하면 오름차순으로 정렬하고, ascending=False로 설정하면 내림차순으로 정렬한다.  
  (기본은 ascending=True 이다.)
 * inplace=False로 설정하면 sort_values()를 호출한 DataFrame은 그대로 유지되며 정렬된 DataFrame을 결과로 반환한다.  
    inplace=True로 설정하면 호출한 DataFrame의 정렬 결과를 그대로 적용한다.  
    (기본은 inplace=False 이다.)  
    
 < titanic_df를 Name 칼럼으로 오름차순 정렬해 반환하기 >
 

In [85]:
titanic_sorted = titanic_df.sort_values(by=['Name'])
titanic_sorted.head(3)

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch     Ticket   Fare Cabin Embarked
845          846         0       3  Abbing, Mr....    male  42.0      0      0  C.A. 5547   7.55   NaN        S
746          747         0       3  Abbott, Mr....    male  16.0      1      1  C.A. 2673  20.25   NaN        S
279          280         1       3  Abbott, Mrs...  female  35.0      1      1  C.A. 2673  20.25   NaN        S

* 여러 개의 칼럼으로 정렬하려면 by에 리스트 형식으로 정렬하려는 칼럼을 입력한다. 

< Pclass와 Name을 내림차순으로 정렬한 결과를 반환하기 >

In [86]:
titanic_sorted = titanic_df.sort_values(by=['Pclass', 'Name'], ascending=False)
titanic_sorted.head(3)

PassengerId  Survived  Pclass            Name   Sex   Age  SibSp  Parch    Ticket  Fare Cabin Embarked
868          869         0       3  van Melkebe...  male   NaN      0      0    345777   9.5   NaN        S
153          154         0       3  van Billiar...  male  40.5      0      2  A/5. 851  14.5   NaN        S
282          283         0       3  de Pelsmaek...  male  16.0      0      0    345778   9.5   NaN        S

### (2) Aggregation 함수 적용

* DataFrame에서 바로 aggregation 함수를 호출할 경우 모든 칼럼에 해당 aggregation을 적용한다.

In [87]:
titanic_df.count()

PassengerId    891
Survived       891
Pclass         891
Name           891
Sex            891
Age            714
SibSp          891
Parch          891
Ticket         891
Fare           891
Cabin          204
Embarked       889
dtype: int64

* 특정 칼럼에 aggregation 함수를 적용하기 위해서는 DataFrame에 대상 칼럼들만 추출해 aggregation을 적용하면 된다.

In [88]:
titanic_df[['Age', 'Fare']].mean()

Age     29.699118
Fare    32.204208
dtype: float64

### (3) groupby() 적용

* DataFrame의 groupby() 사용 시 입력 파라미터 by에 칼럼을 입력하면 대상 칼럼으로 groupby 된다.
 * DataFrameGroupBy라는 또 다른 형태의 DataFrame을 반환한다.  
 
 
< Pclass 칼럼 기준으로 GroupBy 하기 >


In [89]:
titanic_groupby = titanic_df.groupby('Pclass')
print(type(titanic_groupby))

<class 'pandas.core.groupby.generic.DataFrameGroupBy'>


* DataFrame에 groupby()를 호출해 반환한 결과에 aggregation 함수를 호출하면 groupby() 대상 칼럼을 제외한 모든 칼럼에 해당 aggregation 함수를 적용한다.

In [90]:
titanic_groupby = titanic_df.groupby('Pclass').count()
titanic_groupby

PassengerId  Survived  Name  Sex  Age  SibSp  Parch  Ticket  Fare  Cabin  Embarked
Pclass                                                                                    
1               216       216   216  216  186    216    216     216   216    176       214
2               184       184   184  184  173    184    184     184   184     16       184
3               491       491   491  491  355    491    491     491   491     12       491

* DataFrame의 groupby()에 특정 칼럼만 aggregation 함수를 적용하려면 groupby()로 반환된 DataFrameGroupBy 객체에 해당 칼럼을 필터링한 뒤 aggregation 함수를 적용한다.

In [92]:
titanic_groupby = titanic_df.groupby('Pclass')[['PassengerId', 'Survived']].count()
titanic_groupby

PassengerId  Survived
Pclass                       
1               216       216
2               184       184
3               491       491

* DataFrame groupby()의 경우 적용하려는 여러 개의 aggregation 함수명을 DataFrameGroupBy 객체의 agg() 내에 인자로 입력해서 사용한다.

In [93]:
titanic_df.groupby('Pclass')['Age'].agg([max, min])

max   min
Pclass            
1       80.0  0.92
2       70.0  0.67
3       74.0  0.42

* 여러 개의 칼럼이 서로 다른 aggregation 함수를 groupby에서 호출하려면 agg() 내에 입력 값으로 딕셔너리 형태로 aggregation이 적용될 칼럼들과 aggregation 함수를 입력한다.

In [95]:
agg_format = {'Age':'max', 'SibSp':'sum', 'Fare':'mean'}
titanic_df.groupby('Pclass').agg(agg_format)

Age  SibSp       Fare
Pclass                        
1       80.0     90  84.154687
2       70.0     74  20.662183
3       74.0    302  13.675550

## [8] 결손 데이터 처리하기
---

* 결손 데이터는 칼럼에 값이 없는, 즉 NULL인 경우를 의미하여, 이를 넘파이의 NaN으로 표시한다. 
* NaN 값은 평균, 총합 등의 함수 연산 시 제외된다. 

### (1) isna()로 결손 데이터 여부 확인
* DataFrame에 isna()를 수행하면 모든 칼럼의 값이 NaN인지 아닌지를 True나 False로 알려준다.

In [96]:
titanic_df.isna().head(3)

PassengerId  Survived  Pclass   Name    Sex    Age  SibSp  Parch  Ticket   Fare  Cabin  Embarked
0        False     False   False  False  False  False  False  False   False  False   True     False
1        False     False   False  False  False  False  False  False   False  False  False     False
2        False     False   False  False  False  False  False  False   False  False   True     False

* 결손 데이터의 개수는 isna() 결과에 sum() 함수를 추가해 구할 수 있다.
* sum() 호출 시 True는 내부적으로 숫자 1로, False는 숫자 0으로 변환되므로 결손 데이터의 개수를 구할 수 있다.

In [97]:
titanic_df.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

### (2) fillna()로 결손 데이터 대체하기   

< 타이타닉 데이터 세트의 'Cabin'칼럼의 NaN 값을 'C000'으로 대체하기 >

In [98]:
titanic_df['Cabin'] = titanic_df['Cabin'].fillna('C000')
titanic_df.head(3)

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked
0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500  C000        S
1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C
2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250  C000        S

* fillna()를 이용해 반환 값을 다시 받거나 inplace=True 파라미터를 fillna()에 추가해야 실제 데이터 세트 값이 변경된다.

In [99]:
# 첫 번째 방법
titanic_df['Cabin'] = titanic_df['Cabin'].fillna('C000')
# 두 번째 방법
titanic_df['Cabin'].fillna('C000', inplace=True)

< 'Age'칼럼의 NaN 값을 평균 나이로, 'Embarked'칼럼의 NaN 값을 'S'로 대체해 모든 결손 데이터를 처리하기 >

In [101]:
titanic_df['Age'].fillna(titanic_df['Age'].mean(), inplace=True)
titanic_df['Embarked'].fillna('S', inplace=True)
titanic_df.isna().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       0
dtype: int64

## [9] apply lambda 식으로 데이터 가공
---

판다스는 apply 함수에 lambda 식을 결합해 DataFrame이나 Series의 레코드별로 데이터를 가공하는 기능을 제공한다. 판다스의 경우 칼럼에 일괄적으로 데이터 가공을 하는 것이 속도 면에서 더 빠르나 복잡한 데이터 가공이 필요할 경우 어쩔 수 없이 apply lambda를 이용한다.

In [102]:
def get_square(a):
    return a**2

print(get_square(3))

# lambda 식으로 변환
lambda_square = lambda x : x**2
print(lambda_square(3))

9
9


* labmba x : x ** 2  

 : 왼쪽 x는 입력 인자이고 오른쪽 식은 호출 시 계산 결과가 반환된다.
 
* lambda 식을 이용할 때 여러 개의 값을 입력 인자로 사용해야 할 경우에는 보통 map() 함수를 결합해서 사용한다.

In [103]:
a = [1, 2, 3]
squares = map(lambda x : x**2, a)
list(squares)

[1, 4, 9]

< 'Name' 칼럼의 문자열 개수를 별도의 칼럼인 'Name_len'에 생성하기 >
* DataFrame의 apply에 lambda 식을 적용하기

In [104]:
titanic_df['Name_len'] = titanic_df['Name'].apply(lambda x : len(x))
titanic_df[['Name', 'Name_len']].head(3)

Name  Name_len
0  Braund, Mr....        23
1  Cumings, Mr...        51
2  Heikkinen, ...        22

< 나이가 15세 미만이면 'Child', 그렇지 않으면 'Adult'로 구분하는 새로운 칼럼 'Child_Adult'를 apply lambda를 이용해 만들기>

In [105]:
titanic_df['Child_Adult'] = titanic_df['Age'].apply(lambda x : 'Child' if x < 15 else 'Adult')
titanic_df[['Age', 'Child_Adult']].head(8)

Age Child_Adult
0  22.000000       Adult
1  38.000000       Adult
2  26.000000       Adult
3  35.000000       Adult
4  35.000000       Adult
5  29.699118       Adult
6  54.000000       Adult
7   2.000000       Child

* lambda x : 'Child' if x <= 15 else 'Adult'
 * lambda 식에서 if else를 사용할 경우 if 절의 경우 if 식보다 반환 값을 먼저 기술해야하고, else의 경우는 else 식이 먼저 나오고 반환 값이 나중에 온다.   
 
* 또한, if, else만 지원하고 if, else if, else와 같이 else if는 지원하지 않는다. else if를 이용하기 위해서는 else 절을 ()로 내포해 ()내에서 다시 if else 적용해 사용한다.


< 나이가 15세 이하이면 Child, 15 ~ 60세 사이는 Adult, 61세 이상은 Elderly로 분류하는 'Age_Cat' 칼럼 만들기 >


In [106]:
titanic_df['Age_cat'] = titanic_df['Age'].apply(lambda x : 'Child' if x <= 15 else ('Adult' if x <= 60 else 'Elderly'))
titanic_df['Age_cat'].value_counts()

Adult      786
Child       83
Elderly     22
Name: Age_cat, dtype: int64

< 나이에 따라 더 세분화된 분류하기>

In [107]:
def get_category(age):
    cat = ''
    if age <= 5: cat = 'Baby'
    elif age <= 12: cat = 'Child'
    elif age <= 18: cat = 'Teenager'
    elif age <= 25: cat = 'Student'
    elif age <= 35: cat = 'Young Adult'
    elif age <= 60: cat = 'Adult'
    else : cat = 'Elderly'
        
    return cat

In [108]:
# lambda 식에서 위에서 생성한 get_category() 함수를 반환값으로 지정
titanic_df['Age_cat'] = titanic_df['Age'].apply(lambda x : get_category(x))
titanic_df[['Age', 'Age_cat']].head()

Age      Age_cat
0  22.0      Student
1  38.0        Adult
2  26.0  Young Adult
3  35.0  Young Adult
4  35.0  Young Adult